examples/Custom_Hooks/probability_of_improvement.py

```diff
@@ -0,0 +1,219 @@
+        poly = PolyCollection(verts, color=color, alpha=0.9)
+        ax.add_collection3d(poly, zs=i, zdir="x")
+
+        ax.plot(x, y, z, zdir="x", color=color)
+
+    # Set viewing angle
+    ax.view_init(elev=20, azim=30)
+
+    # Reduce space between the iterations
+    ax.set_box_aspect([0.7, 1, 1])
+
+    # Set the y-axis limit based on the maximal PI
+    ax.set_ylim(0, pi_max)
+
+    # Set x-axis ticks to have the correct iteration number
+    ax.set_xticks(np.arange(0, len(pi_list), 1))
+    ax.set_xticklabels([i for i in range(1, len(pi_list) + 1)])
+
+    ax.set_ylabel("PI")
+    ax.set_xlabel("Iteration")
+    ax.set_zlabel("Density")
+    ax.set_title("Investigate the PI")
+
+    output_path = Path(path, base_name)
+    # mypy thinks that ax.figure might become None, hence the explicit ignore
+    if isinstance(ax.figure, Figure):
+        ax.figure.savefig(
+            output_path,
+            format="svg",
+            transparent=True,
+        )
+    else:
+        warnings.warn("Plots could not be saved.")
+    plt.close()
+    return ax.get_figure()
+
+
+### Monkeypatching
+
+# Next, we create our recommender and monkeypatch its `recommend` method:
+
+my_recommender = TwoPhaseMetaRecommender(
+    initial_recommender=RandomRecommender(),
+    recommender=BotorchRecommender(
+        surrogate_model=GaussianProcessSurrogate(),
+        allow_repeated_recommendations=True,
+        allow_recommending_already_measured=True,
+    ),
+)
+my_recommender.recommender.recommend = MethodType(
+    register_hooks(
+        BotorchRecommender.recommend,
+        post_hooks=[check_probability_of_improvement],
+    ),
+    my_recommender.recommender,
+)
+
+# In this example we use `MethodType` to bind the `RandomRecommender.recommend`
+# **function** with our hook.
+# For more information, we refer to the [`basics example`](./basics.md).
+
+
+### Triggering the Hook
+
+# We setup the other objects to trigger the hook:
+
+test_function = Hartmann(dim=DIMENSION)
+
+WRAPPED_FUNCTION = botorch_function_wrapper(test_function=test_function)
+
+BOUNDS = test_function.bounds
+
+discrete_params = [
+    NumericalDiscreteParameter(
+        name=f"x{d}",
+        values=np.linspace(lower, upper, POINTS_PER_DIM),
+    )
+    for d, (lower, upper) in enumerate(BOUNDS.T)
+]
+
+searchspace = SearchSpace.from_product(parameters=discrete_params)
+objective = SingleTargetObjective(target=NumericalTarget(name="Target", mode="MIN"))
+campaign = Campaign(
+    searchspace=searchspace,
+    recommender=my_recommender,
+    objective=objective,
+)
+
+# Now we perform 7 recommendations:
+
+for i in range(N_DOE_ITERATIONS):
+    recommendation = campaign.recommend(BATCH_SIZE)
+    target_values = recommendation.apply(WRAPPED_FUNCTION, axis=1)
+    recommendation["Target"] = target_values
+    campaign.add_measurements(recommendation)
+
+# Lastly, we plot the PI from the previous iterations to be able to analyse them:
```

Comment:
You actually don't analyse them in the example. So unless you add some interpreting comment afterward, this is obsolete